### DISCUSSION TITLE
ext-session-lock script that I use

### DISCUSSION THREAD

**Discussion by FI3EI2** [Aylur/ags#291] [Ideas]
Hello,

I've made a script for using ext-session-lock on lockscreen window. It's as follows:

```python
#! /usr/bin/env python

import signal
from pywayland.client import Display
from pywayland.protocol.ext_session_lock_v1 import ExtSessionLockManagerV1

def handler(reg, pid, iface, ver):
    data = reg.user_data
    if iface == "ext_session_lock_manager_v1":
       data["locker"] = reg.bind(pid, ExtSessionLockManagerV1, ver)
    return 1

def remover(reg, pid):
    # Use if handler remover needed
    pass

def dispatcher(proxy):
    # Use if handler locked or finished event listeners needed
    pass

data = {}
display = Display()
display.connect()

registry = display.get_registry()
registry.dispatcher["global"] = handler
registry.dispatcher["global_remove"] = remover
registry.user_data = data

display.dispatch(block=True)
display.roundtrip()

locker = data["locker"].lock()
locker.dispatcher["locked"] = dispatcher
locker.dispatcher["finished"] = dispatcher

def unlocker(sig=None, frame=None):
    locker.unlock_and_destroy()
    display.disconnect()
    
signal.signal(signal.SIGUSR1, unlocker)
signal.signal(signal.SIGINT, unlocker)

display.dispatch(block=True)
display.roundtrip()

signal.pause()
```

And then, I've made some additions to Aylur's lockscreen config:
```javascript
import App from 'resource:///com/github/Aylur/ags/app.js';
import Service from 'resource:///com/github/Aylur/ags/service.js';
import * as Utils from 'resource:///com/github/Aylur/ags/utils.js';

const auth = App.configDir + '/services/lock/auth.py';
const lock = App.configDir + '/services/lock/lock.py';

class Lockscreen extends Service {
    static {
        Service.register(this, {
            'lock': ['boolean'],
            'authenticating': ['boolean'],
        });
    }

    constructor() {
        super(...arguments);
        this.locker = undefined;
    }

    auth(password) {
        this.emit('authenticating', true);
        Utils.execAsync([auth, password])
            .then(out => {
                this.emit('lock', out !== 'True');
                this.emit('authenticating', false);
                if (this.locker) this.locker.force_exit();
            }).catch(err => console.error(err));
    }

    lockscreen() {
        this.emit('lock', true);
        this.locker = Utils.subprocess(lock, () => {}, (err) => console.error(err));
    };
}

export default new Lockscreen();
```

Here's how it works;
If you activate the lockscreen, the screen will be shown and session will be locked. When you make a success auth, then screen will disappear and session lock will be removed.

Simple is that :)

**Comment by kotontrion** [comment]
This should not work. From the ext-session-lock protocol doc:

>  The locked event indicates that the session is locked. This means that the compositor must stop rendering and providing input to normal clients. Instead the compositor must blank all outputs with an opaque color such that their normal content is fully hidden.
> 
> The only surfaces that should be rendered while the session is locked are the lock surfaces created through this interface and optionally, at the compositor's discretion, special privileged surfaces such as input methods or portions of desktop shell UIs.

So I would expect running this would result in a black screen.

I tested this and it seems to work with Hyprland, so i assume this protocol is not correctly implemented there. On sway i get the expected black screen.

**Reply by FI3EI2** [reply]
Yes, I'm using Hyprland, and it looks like this.
![image](https://github.com/Aylur/ags/assets/33766842/2b79680b-b997-42d6-a055-2c4547b84bf2)

I can't click icons or use keybindings. But I can use mouse scroll up and down.

Anyway, I made another script, but it uses swaylock, it'll probably work on sway, but it's not a general solution anyway.

For using ext-session-lock, you must give a wlSurface and wlOutput parameters for getting lock surface. I didn't find a way to render ags as a wlSurface, then I made this. As I said, it works hyprland, swaylock could be use with the same idea too.

**Reply by kotontrion** [reply]
Yes, I know about these issues with wl_surface, gtklock has the same problem.

I just wanted to point out that this works due to improper implementation of the protocol and therefore doesn't work with all compositors and with Hyprland it could stop working at any moment if they decide to fix this unintended behavior.

**Reply by FI3EI2** [reply]
Yes, you're right. I want to ask a question to you, if you have sway wm, because I don't have sway. When you use swaylock on that, can you make the background transparent with `swaylock -c 00000000` and can see through the lock surface?

**Reply by kotontrion** [reply]
No you can't. sway will set the background to a black screen, so making swaylock transparent will just show sways black screen. The intention is that the lockscreen hides the content of the screen, which could be sensitive. As a workaround you could make a screenshot and set it as a background image for swaylock.

**Reply by FI3EI2** [reply]
Ok, believe or not, I don't want Hyprland to fully implement it, maybe want wayland devs to change that policy :) I'm using xray on for my lock layer, and it hides every client below the overlay layer, and make my animated wallpaper visible. So I make my sensitive screens invisible anyway. But before either Hyprland making the appropriate changes or our layer works on session lock, one can use this script, that's all I can do for now :)

**Comment by kotontrion** [comment]
lockscreens can now be properly implemened in ags using this: [gtk-session-lock](https://github.com/Cu3PO42/gtk-session-lock) (gtk-session-lock-git in AUR)
I already made one [in my dotfiles](https://github.com/kotontrion/dotfiles/blob/main/.config/ags/lockscreen.js)
generally like this: 
```js
import Lock from "gi://GtkSessionLock";
import Gdk from "gi://Gdk?version=3.0";
import Gtk from "gi://Gtk?version=3.0";

//init session lock
const lock = Lock.prepare_lock();

//lock
lock.lock_lock();

//create a window for each monitor
const display = Gdk.Display.get_default();
for (let m = 0;  m < display?.get_n_monitors();  m++) {
  const monitor = display?.get_monitor(m);
  const win = new Gtk.Window({ ... });
  lock.new_surface(win, monitor);
  win.show_all();
}

//to unlock
lock.unlock_and_destroy();
//only if ags should be terminated after unlock
Gdk.Display.get_default()?.sync();
App.quit()

```
